In [ ]:
import pandas as pd

: 

In [2]:
data = pd.read_csv('NeedsWrangling.csv')

In [3]:
data

CustomerID  Firstname  Lastname          Address      City State    Zip   
0            1      Billy     Davis      121 Main St  Prescott    AZ  86305  \
1            2      Betty     Davis  121 Main Street  Prescott    AZ  86305   
2            3    William  Franklin    120 Miller St  Prescott    AZ  86303   
3            4       John     Smith     3331 Main St  prescott    az  86302   
4            5      Besty    Watson         121 Main  Prescott    Az  86305   
5            6      Louis     Davis     121 Main Str  Prescott    az   8601   
6            7      Oscar     Swane      333 Main St  Prescott    AZ  86302   
7            8      Allen    Lawson     555 Penelope  Prescott    AZ  86305   
8            1      Billy     Davis      121 Main St  Prescott    AZ  86305   
9            6      Louis     Davis     121 Main Str  Prescott    az   8601   
10          11     Laurie   Wallace         1 Gatsby  Prescott    AZ  86305   
11          12    William    Frank      5 West Smith  Prescutt    AZ  86305   

          Email  
0    Bd@dma.net  
1   db2@dma.net  
2    wf@dma.net  
3    js@dma.net  
4    bw@dma.net  
5    ld@dma.net  
6    os@dma.net  
7    al@dma.net  
8    Bd@dma.net  
9    ld@dma.net  
10   lw@dma.net  
11   wf@dma.net

In [4]:
data.head()

CustomerID Firstname  Lastname          Address      City State    Zip   
0           1     Billy     Davis      121 Main St  Prescott    AZ  86305  \
1           2     Betty     Davis  121 Main Street  Prescott    AZ  86305   
2           3   William  Franklin    120 Miller St  Prescott    AZ  86303   
3           4      John     Smith     3331 Main St  prescott    az  86302   
4           5     Besty    Watson         121 Main  Prescott    Az  86305   

         Email  
0   Bd@dma.net  
1  db2@dma.net  
2   wf@dma.net  
3   js@dma.net  
4   bw@dma.net

In [5]:
data.duplicated()

0     False
1     False
2     False
3     False
4     False
5     False
6     False
7     False
8      True
9      True
10    False
11    False
dtype: bool

In [6]:
data2 = data.drop_duplicates('CustomerID')

In [7]:
data2

CustomerID  Firstname  Lastname          Address      City State    Zip   
0            1      Billy     Davis      121 Main St  Prescott    AZ  86305  \
1            2      Betty     Davis  121 Main Street  Prescott    AZ  86305   
2            3    William  Franklin    120 Miller St  Prescott    AZ  86303   
3            4       John     Smith     3331 Main St  prescott    az  86302   
4            5      Besty    Watson         121 Main  Prescott    Az  86305   
5            6      Louis     Davis     121 Main Str  Prescott    az   8601   
6            7      Oscar     Swane      333 Main St  Prescott    AZ  86302   
7            8      Allen    Lawson     555 Penelope  Prescott    AZ  86305   
10          11     Laurie   Wallace         1 Gatsby  Prescott    AZ  86305   
11          12    William    Frank      5 West Smith  Prescutt    AZ  86305   

          Email  
0    Bd@dma.net  
1   db2@dma.net  
2    wf@dma.net  
3    js@dma.net  
4    bw@dma.net  
5    ld@dma.net  
6    os@dma.net  
7    al@dma.net  
10   lw@dma.net  
11   wf@dma.net

In [8]:
data2['Length'] = data2['Firstname'].map(len)

C:\Users\zeeha\AppData\Local\Temp\ipykernel_26552\1050047626.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data2['Length'] = data2['Firstname'].map(len)


In [9]:
data2.loc[:,'Length'] = data2['Firstname'].map(len)

In [10]:
data2

CustomerID  Firstname  Lastname          Address      City State    Zip   
0            1      Billy     Davis      121 Main St  Prescott    AZ  86305  \
1            2      Betty     Davis  121 Main Street  Prescott    AZ  86305   
2            3    William  Franklin    120 Miller St  Prescott    AZ  86303   
3            4       John     Smith     3331 Main St  prescott    az  86302   
4            5      Besty    Watson         121 Main  Prescott    Az  86305   
5            6      Louis     Davis     121 Main Str  Prescott    az   8601   
6            7      Oscar     Swane      333 Main St  Prescott    AZ  86302   
7            8      Allen    Lawson     555 Penelope  Prescott    AZ  86305   
10          11     Laurie   Wallace         1 Gatsby  Prescott    AZ  86305   
11          12    William    Frank      5 West Smith  Prescutt    AZ  86305   

          Email  Length  
0    Bd@dma.net       5  
1   db2@dma.net       5  
2    wf@dma.net       7  
3    js@dma.net       4  
4    bw@dma.net       5  
5    ld@dma.net       5  
6    os@dma.net       5  
7    al@dma.net       5  
10   lw@dma.net       6  
11   wf@dma.net       9

In [11]:
data2['Firstname'][2]

'William'

In [12]:
len(data2['Firstname'][2])

7

In [13]:
data2['Firstname'][11]

'  William'

In [14]:
len(data2['Firstname'][11])

9

In [15]:
# data2['Firstname'] = data2['Firstname'].map(str.strip)
### OR
# data2['Firstname'] = data2['Firstname'].str.strip()
data2.loc[:,'Firstname'] = data2['Firstname'].str.strip()

In [16]:
data2['Firstname'][2]

'William'

In [17]:
data2['Firstname'][11]

'William'

In [18]:
data2

CustomerID Firstname  Lastname          Address      City State    Zip   
0            1     Billy     Davis      121 Main St  Prescott    AZ  86305  \
1            2     Betty     Davis  121 Main Street  Prescott    AZ  86305   
2            3   William  Franklin    120 Miller St  Prescott    AZ  86303   
3            4      John     Smith     3331 Main St  prescott    az  86302   
4            5     Besty    Watson         121 Main  Prescott    Az  86305   
5            6     Louis     Davis     121 Main Str  Prescott    az   8601   
6            7     Oscar     Swane      333 Main St  Prescott    AZ  86302   
7            8     Allen    Lawson     555 Penelope  Prescott    AZ  86305   
10          11    Laurie   Wallace         1 Gatsby  Prescott    AZ  86305   
11          12   William    Frank      5 West Smith  Prescutt    AZ  86305   

          Email  Length  
0    Bd@dma.net       5  
1   db2@dma.net       5  
2    wf@dma.net       7  
3    js@dma.net       4  
4    bw@dma.net       5  
5    ld@dma.net       5  
6    os@dma.net       5  
7    al@dma.net       5  
10   lw@dma.net       6  
11   wf@dma.net       9

In [19]:
data2.loc[:,'Length'] = data2['Firstname'].map(len)

In [20]:
data2

CustomerID Firstname  Lastname          Address      City State    Zip   
0            1     Billy     Davis      121 Main St  Prescott    AZ  86305  \
1            2     Betty     Davis  121 Main Street  Prescott    AZ  86305   
2            3   William  Franklin    120 Miller St  Prescott    AZ  86303   
3            4      John     Smith     3331 Main St  prescott    az  86302   
4            5     Besty    Watson         121 Main  Prescott    Az  86305   
5            6     Louis     Davis     121 Main Str  Prescott    az   8601   
6            7     Oscar     Swane      333 Main St  Prescott    AZ  86302   
7            8     Allen    Lawson     555 Penelope  Prescott    AZ  86305   
10          11    Laurie   Wallace         1 Gatsby  Prescott    AZ  86305   
11          12   William    Frank      5 West Smith  Prescutt    AZ  86305   

          Email  Length  
0    Bd@dma.net       5  
1   db2@dma.net       5  
2    wf@dma.net       7  
3    js@dma.net       4  
4    bw@dma.net       5  
5    ld@dma.net       5  
6    os@dma.net       5  
7    al@dma.net       5  
10   lw@dma.net       6  
11   wf@dma.net       7

In [21]:
# data2['State'] = data2['State'].str.upper()
data2.loc[:,'State'] = data2['State'].str.upper()

In [22]:
data2

CustomerID Firstname  Lastname          Address      City State    Zip   
0            1     Billy     Davis      121 Main St  Prescott    AZ  86305  \
1            2     Betty     Davis  121 Main Street  Prescott    AZ  86305   
2            3   William  Franklin    120 Miller St  Prescott    AZ  86303   
3            4      John     Smith     3331 Main St  prescott    AZ  86302   
4            5     Besty    Watson         121 Main  Prescott    AZ  86305   
5            6     Louis     Davis     121 Main Str  Prescott    AZ   8601   
6            7     Oscar     Swane      333 Main St  Prescott    AZ  86302   
7            8     Allen    Lawson     555 Penelope  Prescott    AZ  86305   
10          11    Laurie   Wallace         1 Gatsby  Prescott    AZ  86305   
11          12   William    Frank      5 West Smith  Prescutt    AZ  86305   

          Email  Length  
0    Bd@dma.net       5  
1   db2@dma.net       5  
2    wf@dma.net       7  
3    js@dma.net       4  
4    bw@dma.net       5  
5    ld@dma.net       5  
6    os@dma.net       5  
7    al@dma.net       5  
10   lw@dma.net       6  
11   wf@dma.net       7

In [23]:
data2['Address'].str.contains('121 Main St')

0      True
1      True
2     False
3     False
4     False
5      True
6     False
7     False
10    False
11    False
Name: Address, dtype: bool

In [24]:
idx = data2['Address'].str.contains('121 Main St')

In [25]:
data2.loc[idx,'Address'] = '121 Main St'

In [26]:
data2

CustomerID Firstname  Lastname          Address      City State    Zip   
0            1     Billy     Davis      121 Main St  Prescott    AZ  86305  \
1            2     Betty     Davis      121 Main St  Prescott    AZ  86305   
2            3   William  Franklin    120 Miller St  Prescott    AZ  86303   
3            4      John     Smith     3331 Main St  prescott    AZ  86302   
4            5     Besty    Watson         121 Main  Prescott    AZ  86305   
5            6     Louis     Davis      121 Main St  Prescott    AZ   8601   
6            7     Oscar     Swane      333 Main St  Prescott    AZ  86302   
7            8     Allen    Lawson     555 Penelope  Prescott    AZ  86305   
10          11    Laurie   Wallace         1 Gatsby  Prescott    AZ  86305   
11          12   William    Frank      5 West Smith  Prescutt    AZ  86305   

          Email  Length  
0    Bd@dma.net       5  
1   db2@dma.net       5  
2    wf@dma.net       7  
3    js@dma.net       4  
4    bw@dma.net       5  
5    ld@dma.net       5  
6    os@dma.net       5  
7    al@dma.net       5  
10   lw@dma.net       6  
11   wf@dma.net       7

In [27]:
data2['Address'].value_counts()

Address
121 Main St        3
120 Miller St      1
3331 Main St       1
121 Main           1
333 Main St        1
555 Penelope       1
1 Gatsby           1
   5 West Smith    1
Name: count, dtype: int64

In [28]:
data3 = pd.read_csv('NeedsWranglingJoin.csv')

In [29]:
data3.head()

CustomerID Gender
0           1   Male
1           3   Male
2           4   Male
3           6   Male
4           7   Male

In [30]:
join_data = pd.merge(data2,data3,on='CustomerID')

In [31]:
join_data

CustomerID Firstname  Lastname          Address      City State    Zip   
0            1     Billy     Davis      121 Main St  Prescott    AZ  86305  \
1            1     Billy     Davis      121 Main St  Prescott    AZ  86305   
2            2     Betty     Davis      121 Main St  Prescott    AZ  86305   
3            3   William  Franklin    120 Miller St  Prescott    AZ  86303   
4            4      John     Smith     3331 Main St  prescott    AZ  86302   
5            5     Besty    Watson         121 Main  Prescott    AZ  86305   
6            6     Louis     Davis      121 Main St  Prescott    AZ   8601   
7            6     Louis     Davis      121 Main St  Prescott    AZ   8601   
8            7     Oscar     Swane      333 Main St  Prescott    AZ  86302   
9            8     Allen    Lawson     555 Penelope  Prescott    AZ  86305   
10          11    Laurie   Wallace         1 Gatsby  Prescott    AZ  86305   
11          12   William    Frank      5 West Smith  Prescutt    AZ  86305   

          Email  Length  Gender  
0    Bd@dma.net       5    Male  
1    Bd@dma.net       5    Male  
2   db2@dma.net       5  Female  
3    wf@dma.net       7    Male  
4    js@dma.net       4    Male  
5    bw@dma.net       5  Female  
6    ld@dma.net       5    Male  
7    ld@dma.net       5    Male  
8    os@dma.net       5    Male  
9    al@dma.net       5    Male  
10   lw@dma.net       6  Female  
11   wf@dma.net       7    Male

# CS 457 - Homework Assignment 2: Data Wrangling
## Name: Owais Waheed (ow07611)
### Email: ow07611@st.habib.edu.pk


## Part 1 (40 Points)

In [225]:
import pandas as pd
import numpy as np

In [226]:
df = pd.read_csv('580SurveyCleanup.csv')

In [227]:
df.head()

Unnamed: 0       Unnamed: 1 Unnamed: 2          Unnamed: 3   Unnamed: 4  \
0        No.             Time     Status  Duration (Seconds)  Access Code   
1          1  10/3/20 5:49 PM  Completed               11646          NaN   
2          2  10/3/20 5:49 PM  Completed                 113          NaN   
3          3  10/3/20 5:49 PM  Completed                 572          NaN   
4          4  10/3/20 5:50 PM  Completed                 365          NaN   

      Unnamed: 5      Unnamed: 6     Unnamed: 7 Unnamed: 8  \
0  Email Address      IP Address        Country     Region   
1            NaN  100.36.167.226  United States   Virginia   
2            NaN   98.169.26.134  United States   Virginia   
3            NaN    98.169.52.23  United States   Virginia   
4            NaN  69.255.182.253  United States   Virginia   

               Unnamed: 9  ...             Q11.1         Q11.2  \
0  City / Town / District  ...  Some familiarity  Average user   
1                 Fairfax  ...               NaN           NaN   
2                 Fairfax  ...               NaN             1   
3          Greenway Downs  ...               NaN           NaN   
4             Gainesville  ...               NaN           NaN   

                       Q11.3          Q11.4          Q12             Q12.1  \
0  Frequent use for projects  Fluent/expert  Little/none  Some familiarity   
1                        NaN            NaN          NaN                 1   
2                        NaN            NaN            1               NaN   
3                        NaN            NaN          NaN               NaN   
4                          1            NaN          NaN                 1   

          Q12.2                      Q12.3          Q12.4  \
0  Average user  Frequent use for projects  Fluent/expert   
1           NaN                        NaN            NaN   
2           NaN                        NaN            NaN   
3             1                        NaN            NaN   
4           NaN                        NaN            NaN   

                                                 Q13  
0  What is your goal for learning about data anal...  
1                                                NaN  
2                                                NaN  
3  As part of my work, I have access to a lot of ...  
4                                                NaN  

[5 rows x 46 columns]

In [228]:
df_1 = df.iloc[:,1:12]

In [229]:
df_1.head()

Unnamed: 1 Unnamed: 2          Unnamed: 3   Unnamed: 4     Unnamed: 5  \
0             Time     Status  Duration (Seconds)  Access Code  Email Address   
1  10/3/20 5:49 PM  Completed               11646          NaN            NaN   
2  10/3/20 5:49 PM  Completed                 113          NaN            NaN   
3  10/3/20 5:49 PM  Completed                 572          NaN            NaN   
4  10/3/20 5:50 PM  Completed                 365          NaN            NaN   

       Unnamed: 6     Unnamed: 7 Unnamed: 8              Unnamed: 9  \
0      IP Address        Country     Region  City / Town / District   
1  100.36.167.226  United States   Virginia                 Fairfax   
2   98.169.26.134  United States   Virginia                 Fairfax   
3    98.169.52.23  United States   Virginia          Greenway Downs   
4  69.255.182.253  United States   Virginia             Gainesville   

  Unnamed: 10 Unnamed: 11  
0    Latitude   Longitude  
1    -77.2891     38.8209  
2     -77.263     38.8605  
3    -77.2331     38.9645  
4    -77.6308     38.8175

### Replacing header with first row

In [230]:
df_1.columns

Index(['Unnamed: 1', 'Unnamed: 2', 'Unnamed: 3', 'Unnamed: 4', 'Unnamed: 5',
       'Unnamed: 6', 'Unnamed: 7', 'Unnamed: 8', 'Unnamed: 9', 'Unnamed: 10',
       'Unnamed: 11'],
      dtype='object')

In [231]:
df_1.columns = df_1.iloc[0]
df_1 = df_1[1:]

In [232]:
df_1.head()

0             Time     Status Duration (Seconds) Access Code Email Address  \
1  10/3/20 5:49 PM  Completed              11646         NaN           NaN   
2  10/3/20 5:49 PM  Completed                113         NaN           NaN   
3  10/3/20 5:49 PM  Completed                572         NaN           NaN   
4  10/3/20 5:50 PM  Completed                365         NaN           NaN   
5  10/3/20 5:51 PM  Completed                 92         NaN           NaN   

0      IP Address        Country    Region City / Town / District  Latitude  \
1  100.36.167.226  United States  Virginia                Fairfax  -77.2891   
2   98.169.26.134  United States  Virginia                Fairfax   -77.263   
3    98.169.52.23  United States  Virginia         Greenway Downs  -77.2331   
4  69.255.182.253  United States  Virginia            Gainesville  -77.6308   
5  138.88.108.230  United States       NaN                    NaN   -97.822   

0 Longitude  
1   38.8209  
2   38.8605  
3   38.9645  
4   38.8175  
5    37.751

### Getting Question 1 coulumns

In [233]:
df2 = df.iloc[:,[12,13]]

In [234]:
df2.head()

Q1  Q1.1
0  2.0   4.0
1  NaN   1.0
2  1.0   NaN
3  1.0   NaN
4  1.0   NaN

### Replacing header with first row

In [235]:
df2.columns = df2.iloc[0].astype(str)
df2 = df2[1:]

In [236]:
df2.head()

0  2.0  4.0
1  NaN  1.0
2  1.0  NaN
3  1.0  NaN
4  1.0  NaN
5  1.0  NaN

### Replace every "1" value with its respective column name

In [237]:
df3 = df2.loc[:, '2.0':'4.0'].replace(1, pd.Series(df2.columns, df2.columns)).fillna('')

In [238]:
df3

0   2.0  4.0
1        4.0
2   2.0     
3   2.0     
4   2.0     
5   2.0     
6   2.0     
7   2.0     
8   2.0     
9   2.0     
10       4.0
11  2.0     
12       4.0
13  2.0     
14  2.0     
15       4.0
16       4.0
17  2.0     
18  2.0     
19       4.0
20       4.0
21       4.0
22  2.0     
23       4.0
24       4.0
25       4.0
26  2.0     
27  2.0     
28       4.0
29       4.0
30  2.0     
31       4.0
32  2.0     
33  2.0     
34       4.0
35  2.0     
36       4.0

### concatenate Q1 columns into one final column

In [239]:
df4 = pd.DataFrame(df3.iloc[:,0] + df3.iloc[:,1])

In [240]:
df4

0
1   4.0
2   2.0
3   2.0
4   2.0
5   2.0
6   2.0
7   2.0
8   2.0
9   2.0
10  4.0
11  2.0
12  4.0
13  2.0
14  2.0
15  4.0
16  4.0
17  2.0
18  2.0
19  4.0
20  4.0
21  4.0
22  2.0
23  4.0
24  4.0
25  4.0
26  2.0
27  2.0
28  4.0
29  4.0
30  2.0
31  4.0
32  2.0
33  2.0
34  4.0
35  2.0
36  4.0

In [241]:
df4.columns = ['Q1']

In [242]:
df4.head()

Q1
1  4.0
2  2.0
3  2.0
4  2.0
5  2.0

In [243]:
df5 = pd.concat([df_1,df4],axis=1)

In [244]:
df5.head()

Time     Status Duration (Seconds) Access Code Email Address  \
1  10/3/20 5:49 PM  Completed              11646         NaN           NaN   
2  10/3/20 5:49 PM  Completed                113         NaN           NaN   
3  10/3/20 5:49 PM  Completed                572         NaN           NaN   
4  10/3/20 5:50 PM  Completed                365         NaN           NaN   
5  10/3/20 5:51 PM  Completed                 92         NaN           NaN   

       IP Address        Country    Region City / Town / District  Latitude  \
1  100.36.167.226  United States  Virginia                Fairfax  -77.2891   
2   98.169.26.134  United States  Virginia                Fairfax   -77.263   
3    98.169.52.23  United States  Virginia         Greenway Downs  -77.2331   
4  69.255.182.253  United States  Virginia            Gainesville  -77.6308   
5  138.88.108.230  United States       NaN                    NaN   -97.822   

  Longitude   Q1  
1   38.8209  4.0  
2   38.8605  2.0  
3   38.9645  2.0  
4   38.8175  2.0  
5    37.751  2.0

In [245]:
df5

Time      Status Duration (Seconds) Access Code Email Address  \
1    10/3/20 5:49 PM   Completed              11646         NaN           NaN   
2    10/3/20 5:49 PM   Completed                113         NaN           NaN   
3    10/3/20 5:49 PM   Completed                572         NaN           NaN   
4    10/3/20 5:50 PM   Completed                365         NaN           NaN   
5    10/3/20 5:51 PM   Completed                 92         NaN           NaN   
6    10/3/20 5:59 PM  Incomplete                221         NaN           NaN   
7    10/3/20 6:23 PM   Completed             115983         NaN           NaN   
8    10/3/20 7:10 PM  Incomplete                 40         NaN           NaN   
9    10/3/20 7:25 PM   Completed                137         NaN           NaN   
10   10/3/20 7:56 PM  Incomplete                179         NaN           NaN   
11   10/3/20 8:14 PM   Completed                127         NaN           NaN   
12   10/3/20 9:38 PM   Completed                269         NaN           NaN   
13   10/3/20 9:41 PM   Completed                117         NaN           NaN   
14  10/3/20 10:49 PM   Completed               5966         NaN           NaN   
15  10/3/20 11:25 PM   Completed                254         NaN           NaN   
16  10/4/20 12:28 AM   Completed                526         NaN           NaN   
17   10/4/20 5:22 AM   Completed                398         NaN           NaN   
18   10/4/20 5:36 AM   Completed             128253         NaN           NaN   
19   10/4/20 5:47 AM   Completed                239         NaN           NaN   
20  10/4/20 10:32 AM   Completed                255         NaN           NaN   
21  10/4/20 11:03 AM   Completed                256         NaN           NaN   
22  10/4/20 12:34 PM   Completed                674         NaN           NaN   
23   10/4/20 1:36 PM   Completed                130         NaN           NaN   
24   10/4/20 1:50 PM   Completed                275         NaN           NaN   
25   10/4/20 9:28 PM   Completed                325         NaN           NaN   
26   10/5/20 9:37 AM   Completed                257         NaN           NaN   
27   10/5/20 9:36 AM   Completed                657         NaN           NaN   
28  10/5/20 10:36 AM   Completed               1750         NaN           NaN   
29  10/5/20 11:02 AM   Completed                887         NaN           NaN   
30  10/5/20 11:17 AM   Completed                180         NaN           NaN   
31  10/5/20 12:52 PM   Completed                392         NaN           NaN   
32  10/5/20 12:56 PM  Incomplete                192         NaN           NaN   
33   10/5/20 1:52 PM   Completed                277         NaN           NaN   
34   10/5/20 3:34 PM   Completed                119         NaN           NaN   
35   10/5/20 3:41 PM   Completed                149         NaN           NaN   
36   10/5/20 5:58 PM   Completed                285         NaN           NaN   

         IP Address        Country    Region City / Town / District  Latitude  \
1    100.36.167.226  United States  Virginia                Fairfax  -77.2891   
2     98.169.26.134  United States  Virginia                Fairfax   -77.263   
3      98.169.52.23  United States  Virginia         Greenway Downs  -77.2331   
4    69.255.182.253  United States  Virginia            Gainesville  -77.6308   
5    138.88.108.230  United States       NaN                    NaN   -97.822   
6     108.56.134.80  United States  Virginia                 Reston  -77.3489   
7    108.48.175.220  United States  Virginia              Arlington  -77.0987   
8     173.67.199.85  United States  Virginia                Ashburn   -77.539   
9      74.96.114.12  United States  Virginia             Woodbridge  -77.3197   
10    107.77.241.32  United States     Texas                    NaN  -96.8217   
11   138.88.182.245  United States  Maryland            Takoma Park   -77.004   
12     72.83.246.60  United States  Virginia  

#### Formatting for Q.2

In [246]:
dfq2 = df.iloc[:, [14, 15, 16, 17]]

dfq2.columns = dfq2.iloc[0].astype(str)

dfq2 = dfq2[1:]

columns_to_replace = ['Male', 'Female', 'NonBinary', 'Prefer not to answer']

for column in columns_to_replace:
    dfq2[column] = dfq2[column].replace('1', column)
    
dfq2 = dfq2.fillna('')

dfq2['Q2'] = dfq2['Male'] + dfq2['Female'] + dfq2['NonBinary'] + dfq2['Prefer not to answer']

dfq2 = dfq2[['Q2']]

dfq2.describe()



0         Q2
count     36
unique     4
top     Male
freq      18

#### formatting for q8

In [247]:
dfq8 = df.iloc[:,[23,24,25,26]]
dfq8.columns = dfq8.iloc[0].astype(str)

dfq8 = dfq8[1:]

columns_to_replace = ['Microsoft/Windows', 'Apple/MacBook','Other','nan']

for column in columns_to_replace:
    dfq8[column] = dfq8[column].replace(['0','1','2'], column).fillna('')
    
dfq8['Q8'] = dfq8['Microsoft/Windows'] + dfq8['Apple/MacBook'] + dfq8['Other'] + dfq8.iloc[:,3]
    
dfq8 = dfq8[['Q8']]
dfq8



0                               Q8
1                    Apple/MacBook
2                    Apple/MacBook
3                    Apple/MacBook
4                Microsoft/Windows
5                    Apple/MacBook
6                                 
7                Microsoft/Windows
8                                 
9                Microsoft/Windows
10                                
11               Microsoft/Windows
12               Microsoft/Windows
13                   Apple/MacBook
14               Microsoft/Windows
15               Microsoft/Windows
16               Microsoft/Windows
17  Microsoft/WindowsApple/MacBook
18               Microsoft/Windows
19               Microsoft/Windows
20                   Apple/MacBook
21               Microsoft/Windows
22               Microsoft/Windows
23  Microsoft/WindowsApple/MacBook
24               Microsoft/Windows
25               Microsoft/Windows
26               Microsoft/Windows
27                   Apple/MacBook
28               Microsoft/Windows
29               Microsoft/Windows
30                   Apple/MacBook
31               Microsoft/Windows
32  Microsoft/WindowsApple/MacBook
33               Microsoft/Windows
34               Microsoft/Windows
35               Microsoft/Windows
36               OtherAcer/Windows

#### formatting for q9

In [248]:
dfq9 = df.iloc[:,[27,28,29]]
dfq9.columns = dfq9.iloc[0].astype(str)

dfq9 = dfq9[1:]

columns_to_replace = ['Yes, Full Time', 'Working, but not Full Time','Not Working while attending Mason']

for column in columns_to_replace:
    if (column == 'Yes, Full Time'):
        dfq9[column] = dfq9[column].replace('1', 'Yes').fillna('')
    elif(column == 'Not Working while attending Mason'):
        dfq9[column] = dfq9[column].replace('1', 'No').fillna('')
    else:
        dfq9[column] = dfq9[column].replace('1', 'Part Time').fillna('')
    
dfq9['Q9'] = dfq9['Yes, Full Time'] + dfq9['Working, but not Full Time'] + dfq9['Not Working while attending Mason'] 
    
dfq9 = dfq9[['Q9']]
dfq9.head()

0         Q9
1         No
2  Part Time
3        Yes
4         No
5        Yes

#### Formatting for q10

In [249]:
dfq10 = df.iloc[:,[30,31,32,33,34]]
dfq10.columns = dfq10.iloc[0].astype(str)

dfq10 = dfq10[1:]


columns_to_replace = ['Little/none', 'Some familiarity','Average user','Frequent use for projects' , 'Fluent/expert']

for column in columns_to_replace:
    if(column == 'Frequent use for projects' ):
        dfq10[column] = dfq10[column].replace('1', 'Frequent User').fillna('')
    else:
        dfq10[column] = dfq10[column].replace('1', column).fillna('')
    
    
    
dfq10['Q10'] = dfq10.iloc[:,0] + dfq10.iloc[:,1] + dfq10.iloc[:,2] + dfq10.iloc[:,3] + dfq10.iloc[:,4]
    
dfq10 = dfq10[['Q10']]
dfq10.head()



0               Q10
1  Some familiarity
2      Average user
3  Some familiarity
4      Average user
5  Some familiarity

#### formatting for q11

In [250]:
dfq11 = df.iloc[:,[35,36,37,38,39]]
dfq11.columns = dfq11.iloc[0].astype(str)

dfq11 = dfq11[1:]


columns_to_replace = ['Little/none', 'Some familiarity','Average user','Frequent use for projects' , 'Fluent/expert']

for column in columns_to_replace:
    if(column == 'Frequent use for projects' ):
        dfq11[column] = dfq11[column].replace('1', 'Frequent User').fillna('')
    else:
        dfq11[column] = dfq11[column].replace('1', column).fillna('')
    
    
    
dfq11['Q11'] = dfq11.iloc[:,0] + dfq11.iloc[:,1] + dfq11.iloc[:,2] + dfq11.iloc[:,3] + dfq11.iloc[:,4]
    
dfq11 = dfq11[['Q11']]
dfq11.head()

0               Q11
1       Little/none
2      Average user
3       Little/none
4     Frequent User
5  Some familiarity

#### Formatting for q12

In [251]:
dfq12 = df.iloc[:,[40,41,42,43,44]]
dfq12.columns = dfq12.iloc[0].astype(str)

dfq12 = dfq12[1:]


columns_to_replace = ['Little/none', 'Some familiarity','Average user','Frequent use for projects' , 'Fluent/expert']

for column in columns_to_replace:
    if(column == 'Frequent use for projects' ):
        dfq12[column] = dfq12[column].replace('1', 'Frequent User').fillna('')
    else:
        dfq12[column] = dfq12[column].replace('1', column).fillna('')
    
    
    
dfq12['Q12'] = dfq12.iloc[:,0] + dfq12.iloc[:,1] + dfq12.iloc[:,2] + dfq12.iloc[:,3] + dfq12.iloc[:,4]
    
dfq12 = dfq12[['Q12']]
dfq12.head()

0               Q12
1  Some familiarity
2       Little/none
3      Average user
4  Some familiarity
5      Average user

#### Combining all data frames into one output for q.1

In [280]:
dfq3to7 = df.iloc[:,[18,19,20,21,22]]

dfq3to7 = dfq3to7[1:]

dfq13 = df.iloc[:,[45]]
dfq13 = dfq13[1:].fillna('No response')

df_out_p1 = pd.concat([df5,dfq2,dfq3to7,dfq8,dfq9,dfq10,dfq11,dfq12,dfq13],axis=1)
df_out_p1.head()

df_out_p1.to_csv('formatted_580SurveyCleanup.csv', index=False)

PermissionError: [Errno 13] Permission denied: 'formatted_580SurveyCleanup.csv'

## Part 2 (30 Points)

In [ ]:
dfc = pd.read_csv('formatted_580SurveyCleanup.csv')
dfc.isnull().sum() 

The above result shows the number of null values present in each coloumn.
This shows that there are some null values in columns Q4 till Q12 , Region, City/town/district.
Since there are 36 entries and both Access code and email address column has 36 null values I am removing these columns as they can't be used due to absence of values.

In [ ]:
# This command will remove all the empty coloumns in the data.
dfc = dfc.drop(['Access Code', 'Email Address'], axis=1)

dfc.head()

#### Cleaning the data:
Before filling in the missing value, I will first standardize the data by fixing typos, capitalizng all letters,etc. For this I will be looking at the values and try to recognize any errors by sight.


In [253]:
'''In q.3 'date of birth' there are two problem, on index 17 there is '2 3' and then 
on index 32 there is '1993' which can be the birth year. Since the survey was conducted in 2020 
value on index 32 should be 27
'''

dfc.at[32,'Q3'] = 27
dfc.at[17,'Q3'] = 23

# dfc['Q3']


In [254]:
'''In Q.4 column we have following errors:
5.2 inches can not be a person's height so it will be a typo surely. Replacing it with 52
6l and 7O are also typos they should be replaced by 61 and 70 respectively
'''

dfc['Q4'].replace('5.2','52',inplace = True)
dfc['Q4'].replace('7O','70',inplace = True)
dfc['Q4'].replace('6l','61',inplace = True)

# dfc['Q4']

In [255]:
''' In Q.5 we need to use title case for every countries and also remove unecessary white spaces.
For Standardizing data I will replace all instances of USA, US with United States
'''
dfc['Q5'].replace(['USA' , 'US' ],'United States',inplace = True)
dfc['Q5'] = dfc['Q5'].str.strip()
dfc['Q5'] = dfc['Q5'].str.title()

# dfc['Q5']


In [256]:
'''In Q8 I am doing a little formatting where multiple laptops are used
'''

dfc['Q8'].replace('Microsoft/WindowsApple/MacBook' , 'Microsoft/Windows and Apple/MacBook', inplace = True)
dfc['Q8'].replace('OtherAcer/Windows' , 'Acer/Windows',inplace = True)


In [257]:


dfc['Q6'].replace('Economic', 'Economics', inplace = True)
dfc['Q6'].replace(['Cse', 'Computer science engineering ', 'Computer engineering'], 'Computer Science and Engineering',inplace = True)
dfc['Q6'].replace(['Bachelor Of Engineering', 'Engineering '], ' Engineering',inplace = True)
dfc['Q6'].replace('Criminology Law And Soci', 'Criminology law and social sciences',inplace = True)
dfc['Q6'].replace('Bit', 'Business Information Technology',inplace = True)
dfc['Q6'].replace('Ece', 'Electronics And Communication Engineering',inplace = True)
dfc['Q6'].replace('Yes', np.NaN,inplace = True)
dfc['Q6'].replace('B.Tech In Cs', 'Computer Science',inplace = True)
dfc['Q6'].replace('Bs Accounting', 'Accounting',inplace = True)

dfc['Q6'] = dfc['Q6'].str.strip()
dfc['Q6'] = dfc['Q6'].str.title()

dfc['Q6']


0                                     Economics
1                                       Finance
2                         Computer Applications
3                           Business Management
4                                     Economics
5                                     Economics
6                                       Finance
7                                     Economics
8                                     Economics
9                        Economics With Finance
10              Business Information Technology
11                         Computer Engineering
12                                      Biology
13             Computer Science And Engineering
14                                   Management
15                       Economics With Finance
16                                    Economics
17                                  Engineering
18                          Accounting And Isom
19                                   Statistics
20                             Computer 

### Filling missing values

after successfully formatting and standardizing the data, I am going to handle missing values by dropping irrelevant coloumns and filling the data using mean, median and mode to predict the value.

In [258]:
# Finding rows where responses were incomplete
del_row = dfc.index[dfc['Status'] == 'Incomplete']
del_row

Int64Index([5, 7, 9, 31], dtype='int64')

''' In the given data there are 4 incomplete responses.
Out of these 4 responses, two have completed less than 30% of the survey that is 3/13 questions and 4/13 questions.
But I won't be dropping these rows as it might lead to false results.
for example, suppose section 1 has one more student than section 2, if we drop these values than section 2 will have more students

The rest of two has completed more than 50% of the survey so I am also keeping those.
'''




In the following section, I have taken mode of the coloumns Region, city, fluency in programming language, type of laptops, Graduation date, and job status 
because it is non numeric and categorical data and there is no natural or inherent order between the entries.

In [259]:
dfc['Region'].fillna(dfc['Region'].mode()[0], inplace=True)
dfc['City / Town / District'].fillna(dfc['City / Town / District'].mode()[0], inplace=True)


dfc['Q5'].fillna(dfc['Q5'].mode()[0], inplace=True)
dfc['Q6'].fillna(dfc['Q6'].mode()[0], inplace=True)
dfc['Q7'].fillna(dfc['Q7'].mode()[0], inplace=True)
dfc['Q8'].fillna(dfc['Q8'].mode()[0], inplace=True)
dfc['Q9'].fillna(dfc['Q9'].mode()[0], inplace=True)
dfc['Q10'].fillna(dfc['Q10'].mode()[0], inplace=True)
dfc['Q11'].fillna(dfc['Q11'].mode()[0], inplace=True)
dfc['Q12'].fillna(dfc['Q12'].mode()[0], inplace=True)



#dfc['Region'] 
#dfc['City / Town / District']
# dfc['Q5']
# dfc['Q6']
# dfc['Q7']
# dfc['Q8']
# dfc['Q9']
# dfc['Q10']
# dfc['Q11']
# dfc['Q12']

In [260]:
''' IN q.4 there is height in inches so I will be taking mean to predict the value because it is numerical value and mean will not change the central tendency of the data.
'''

dfc['Q4'] = pd.to_numeric(dfc['Q4'])
dfc['Q4'].fillna(dfc['Q4'].mean(), inplace=True)

# dfc['Q4']

In [ ]:
dfc.to_csv('cleaned_580SurveyCleanup.csv', index=False)
dfc.isnull().sum()

## Part 3 (30 Points)

In [262]:
dfs = pd.read_csv('cleaned_580SurveyCleanup.csv')
dfs.head()



Time     Status  Duration (Seconds)      IP Address  \
0  10/3/20 5:49 PM  Completed               11646  100.36.167.226   
1  10/3/20 5:49 PM  Completed                 113   98.169.26.134   
2  10/3/20 5:49 PM  Completed                 572    98.169.52.23   
3  10/3/20 5:50 PM  Completed                 365  69.255.182.253   
4  10/3/20 5:51 PM  Completed                  92  138.88.108.230   

         Country    Region City / Town / District  Latitude  Longitude   Q1  \
0  United States  Virginia                Fairfax  -77.2891    38.8209  4.0   
1  United States  Virginia                Fairfax  -77.2630    38.8605  2.0   
2  United States  Virginia         Greenway Downs  -77.2331    38.9645  2.0   
3  United States  Virginia            Gainesville  -77.6308    38.8175  2.0   
4  United States  Virginia                Fairfax  -97.8220    37.7510  2.0   

   ...    Q4             Q5                     Q6          Q7  \
0  ...  63.0         Taiwan              Economics  2021-12-31   
1  ...  59.0          India                Finance  2021-05-01   
2  ...  70.0          India  Computer Applications  2022-06-01   
3  ...  66.0    South Korea    Business Management  2021-12-31   
4  ...  70.0  United States              Economics  2021-01-01   

                  Q8         Q9               Q10               Q11  \
0      Apple/MacBook         No  Some familiarity       Little/none   
1      Apple/MacBook  Part Time      Average user      Average user   
2      Apple/MacBook        Yes  Some familiarity       Little/none   
3  Microsoft/Windows         No      Average user     Frequent User   
4      Apple/MacBook        Yes  Some familiarity  Some familiarity   

                Q12                                                Q13  
0  Some familiarity                                        No response  
1       Little/none                                        No response  
2      Average user  As part of my work, I have access to a lot of ...  
3  Some familiarity                                        No response  
4      Average user                                              Learn  

[5 rows x 22 columns]

In [263]:
print('Summary : \n',dfs['Q1'].describe(include = 'all'),'\n')
print('Unique value counts\n',dfs['Q1'].value_counts())

Summary : 
 count    36.000000
mean      2.888889
std       1.007905
min       2.000000
25%       2.000000
50%       2.000000
75%       4.000000
max       4.000000
Name: Q1, dtype: float64 

Unique value counts
 2.0    20
4.0    16
Name: Q1, dtype: int64


#### Summary Of Q1

Question 1 does not provide much information as they are just labels of the section. From analysis I can say that more students are enrolled in Section 2.0 of the course.


In [264]:
print('Summary : \n',dfs['Q2'].describe(include = 'all'),'\n')
print('Unique value counts\n',dfs['Q2'].value_counts())

Summary : 
 count       36
unique       4
top       Male
freq        18
Name: Q2, dtype: object 

Unique value counts
 Male                    18
Female                  16
Prefer not to answer     1
NonBinary                1
Name: Q2, dtype: int64


#### Summary Of Q2

Again question 2 contains gender label. from here I can say that males are slightly more than female. There is also non-binary. From this data set I can conclude the class has gender diversity and male/female representations are almost equal.

In [265]:
print('Summary : \n',dfs['Q3'].describe(include = 'all'),'\n')
print('Median\n',dfs['Q3'].median(),'\n')
print('Unique value counts\n',dfs['Q3'].value_counts())

Summary : 
 count    36.000000
mean     27.055556
std       6.484903
min      22.000000
25%      23.000000
50%      25.000000
75%      27.250000
max      55.000000
Name: Q3, dtype: float64 

Median
 25.0 

Unique value counts
 23    8
25    6
24    5
26    3
27    3
40    2
30    2
22    2
28    2
31    1
35    1
55    1
Name: Q3, dtype: int64


#### Summary for q3

This age distribution is quite diverse, with most students in their mid-20s, and a median age of 25. There are some interesting outliers like a 55-year-old student.The most common age is 23, with 8 students in that group. This class represents a wide range of ages, making it a diverse and dynamic learning environment.

In [266]:
print('Summary : \n',dfs['Q4'].describe(include = 'all'),'\n')
print('Median\n',dfs['Q4'].median(),'\n')
print('Unique value counts\n',dfs['Q4'].value_counts())

Summary : 
 count    36.000000
mean     67.057143
std       4.863525
min      52.000000
25%      66.000000
50%      67.028571
75%      70.000000
max      75.000000
Name: Q4, dtype: float64 

Median
 67.02857142857144 

Unique value counts
 70.000000    5
66.000000    5
67.000000    5
71.000000    3
68.000000    3
72.000000    2
74.000000    1
56.000000    1
64.000000    1
61.000000    1
75.000000    1
65.000000    1
63.000000    1
60.000000    1
59.000000    1
69.000000    1
52.000000    1
67.057143    1
73.000000    1
Name: Q4, dtype: int64


#### Summary for q4

The height data suggests a fairly uniform distribution, with an average height of around 67 inches. The range spans from 52 to 75 inches, with a median of roughly 67.03 inches, close to the mean.
However, this data does not give much insight related to the course.

In [267]:
print('Summary : \n',dfs['Q5'].describe(include = 'all'),'\n')

print('Unique value counts\n',dfs['Q5'].value_counts())

Summary : 
 count        36
unique       12
top       India
freq         12
Name: Q5, dtype: object 

Unique value counts
 India                       12
United States                9
Uzbekistan                   3
Taiwan                       2
China                        2
Nigeria                      2
South Korea                  1
Pakistan                     1
Romania                      1
Ethiopia                     1
United States ; Pakistan     1
Vietnam                      1
Name: Q5, dtype: int64


#### Summary for Q5

This citizenship data shows a great national diversity in the class with students from different regions of Asia, north africa and eastern europe. One intresting fact is that the highest enrollment students are Indian citizen which indicates that Indians are more inclined 
towards this field. The second highest representation is from the host country. We can say that this university facilitates international students that's why there are way more international students enrolled in the course.


In [268]:
print('Summary : \n',dfs['Q6'].describe(include = 'all'),'\n')
print('Unique value counts\n',dfs['Q6'].value_counts())

Summary : 
 count            36
unique           22
top       Economics
freq              8
Name: Q6, dtype: object 

Unique value counts
 Economics                                    8
Computer Science                             3
Engineering                                  3
Finance                                      3
Economics With Finance                       2
Criminology Law And Social Sciences          1
Accounting                                   1
Quantity Surveying                           1
Electronics And Communication Engineering    1
Civil Engineering                            1
Systems Engineering                          1
Mathematics & Statistics                     1
Accounting And Isom                          1
Statistics                                   1
Management                                   1
Computer Science And Engineering             1
Biology                                      1
Computer Engineering                         1
Business Inform

#### Summary for q6

The dataset reveals a diverse range of undergraduate degrees among students, with Economics and Computer Science being the most prevalent. The majority come from management and science backgrounds,
with a smaller representation from social sciences. This diversity suggests that the course is popular among students from various educational domains and it also shows the potential this field has.

In [269]:
print('Summary : \n',dfs['Q8'].describe(include = 'all'),'\n')
print('Unique value counts\n',dfs['Q8'].value_counts())

Summary : 
 count                    36
unique                    4
top       Microsoft/Windows
freq                     24
Name: Q8, dtype: object 

Unique value counts
 Microsoft/Windows                      24
Apple/MacBook                           8
Microsoft/Windows and Apple/MacBook     3
Acer/Windows                            1
Name: Q8, dtype: int64


#### Summary for Q8
Among students, the choice of laptop platform is diverse but they prefer Windows (25 students), 
followed by Apple(8 students). There are very few which uses both. This shows people still prefer microsoft laptops.
One certain reason could be windows laptop are cheaper as compared to apple ones.

In [270]:
print('Summary : \n',dfs['Q9'].describe(include = 'all'),'\n')
print('Unique value counts\n',dfs['Q9'].value_counts())

Summary : 
 count     36
unique     3
top       No
freq      20
Name: Q9, dtype: object 

Unique value counts
 No           20
Yes          11
Part Time     5
Name: Q9, dtype: int64


#### Summary for q.9

More than half the students are not working which is surprising as they are doing an MS program. 
There are a good number of people doing a job which brings the professional experience in the class.
Since most students are not working, teacher can expect some unprofessional behavior from the class like asking for multiple extensions, etc.


In [271]:
print('Summary : \n',dfs['Q10'].describe(include = 'all'),'\n')
print('Unique value counts\n',dfs['Q10'].value_counts())

Summary : 
 count               36
unique               4
top       Average user
freq                18
Name: Q10, dtype: object 

Unique value counts
 Average user        18
Some familiarity    12
Little/none          4
Frequent User        2
Name: Q10, dtype: int64


#### Summary for q10

It's noticeable that many students have limited experience with Python, especially those from a business management background. This isn't surprising, as data analytics can be technical, 
and Python is a big part of it. Those who've had more exposure are often computer science grads, might find the course more familiar. there are very few who dont know python maybe this includes older people who may be taking this course to switch professions who knows.

In [272]:
print('Summary : \n',dfs['Q11'].describe(include = 'all'),'\n')
print('Unique value counts\n',dfs['Q11'].value_counts())

Summary : 
 count               36
unique               5
top       Average user
freq                12
Name: Q11, dtype: object 

Unique value counts
 Average user        12
Little/none          8
Some familiarity     8
Frequent User        7
Fluent/expert        1
Name: Q11, dtype: int64


#### Summary for Q11
The data suggests that many students have limited experience with R, although a notable group is proficient or frequently use it for projects,
especially compared to Python. This could be because R is taught in business management and computer science programs, which are the majority here.
For the rest, it might be a learning curve, as most haven't used R  before.

In [273]:
print('Summary : \n',dfs['Q12'].describe(include = 'all'),'\n')
print('Unique value counts\n',dfs['Q12'].value_counts())

Summary : 
 count               36
unique               4
top       Average user
freq                18
Name: Q12, dtype: object 

Unique value counts
 Average user        18
Some familiarity    10
Little/none          5
Frequent User        3
Name: Q12, dtype: int64


#### Summary for Q12

Again, SQL is crucial for database management and data analytics. The data shows that many students are either average or 
frequent users of SQL, but there's also a group with limited familiarity. Those who haven't used SQL before might face some initial challenges in the course.

In [274]:

dfc['Q7'] = pd.to_datetime(dfc['Q7'], format='%Y-%m-%d')
print('Summary : \n',dfs['Q7'].describe(include = 'all'),'\n')


dfc_grad_year = dfc['Q7'].dt.year
print('Summary : \n',dfc_grad_year.describe(include = 'all'),'\n')
print('Unique value counts\n',dfc_grad_year.value_counts())

Summary : 
 count             36
unique            24
top       2021-12-20
freq               6
Name: Q7, dtype: object 

Summary : 
 count      36.000000
mean     2021.333333
std         0.534522
min      2021.000000
25%      2021.000000
50%      2021.000000
75%      2022.000000
max      2023.000000
Name: Q7, dtype: float64 

Unique value counts
 2021    25
2022    10
2023     1
Name: Q7, dtype: int64


#### Summary for Q7

The data appears quite diverse, seemingly reflecting random entries from students. 
It doesn't offer extensive insights into the students except that the majority expect to graduate in 2021 and 2022, with a few in 2023.

In [275]:
df5.isnull().sum()

df5['Q1'].value_counts()

2.0    20
4.0    16
Name: Q1, dtype: int64

In [276]:
df5.describe()

Time     Status Duration (Seconds) Access Code  \
count                36         36                 36           0   
unique               34          2                 36           0   
top     10/3/20 5:49 PM  Completed              11646         NaN   
freq                  3         32                  1         NaN   

       Email Address     IP Address        Country    Region  \
count              0             36             36        32   
unique             0             34              1         3   
top              NaN  108.56.134.80  United States  Virginia   
freq             NaN              2             36        25   

       City / Town / District  Latitude Longitude   Q1  
count                      31        36        36   36  
unique                     17        24        24    2  
top                   Fairfax  -77.2866   38.7851  2.0  
freq                        6         4         4   20

In [277]:
df5['Q1'].describe()

count      36
unique      2
top       2.0
freq       20
Name: Q1, dtype: object